### ISSUE TITLE
Hover best practices

### ISSUE THREAD

**Issue by just-boris** [bem/bem-forum-content-ru#252]
Имеется код
```less
.row {
  &__controls { 
  }
  &:hover .row__controls {
  }
}
```
Смущает то, что для row_controls нужно обязательно писать его имя. В результате в соседних правилах описываются стили одного элемента, 
но глазами этого не видно, потому что один раз он пишется как `&_controls`, а второй раз  `.row_controls`.
Как вы пишете такие конструкции?

**Comment by Guria** [comment]
Ну в библиотеке не используются псевдоселекторы, вместо этого в js контролам добавляется модификатор hovered. соответственно стиль будет выглядеть так:
```styl
.row {
  &_controls { 
    &.row_hovered {
    }
  }
}
```

**Comment by just-boris** [comment]
@Guria то есть `&.row_hovered`, соберется в `.row_controls.row_hovered`. Это мне не поможет.

```
.row {
  &_controls { 
     opacity: 0;
  }
  &:hover .row_controls {
     opacity: 1;
  }
}
```

Я хочу чтобы controls показывались, когда курсор оказывается над блоком. В вашем коде controls не покажутся, потому что над невидимыми объектами не бывает ховера

**Comment by Guria** [comment]
Пардон.
Тогда
```styl
.row {
  &_controls { 
    .row_hovered & {
    }
  }
}
```
PS. у вас элемент `controls` с одним подчёркиванием или это модификатор всё же?

**Comment by just-boris** [comment]
@Guria да, controls – это элемент. Исправил подчеркивания для понятности.

Решение ок, но теперь повторяется `.row`.

**Comment by sipayRT** [comment]
@just-boris ну так это вполне норм ;)

**Comment by Guria** [comment]
Нет ничего страшно в повторениях

**Comment by h4** [comment]
Можно использовать стайлус на полную катушку.

```stylus
$ctx = row

.{$ctx} {
  &__controls {
    visibility hidden
    /.{$ctx}:hover {selector()} {
        visibility visible
    }
  }
}
```

**Comment by Guria** [comment]
@h4 можно, но зачем?

**Comment by h4** [comment]
@Guria чтобы всё, что касается стилей элемента `__controls` лежало в одном визуальном блоке.

Вопрос @just-boris был «Как вы пишете такие конструкции?». Мы пишем из так.

**Comment by sipayRT** [comment]
@h4 если уж оптимизировать, то в `$ctx` сразу и класс записывать - `$ctx = '.row'`. Не нужно будет потом везде точку дописывать ;)
Если уйти от примера, то вместо `.row` там может быть что-то и помассивнее, поэтому вариант с {$ctx} клевый. Просто не нужно им злоупотреблять

**Comment by h4** [comment]
@sipayRT точку мы не дописываем, потому что такой ход позволяет делать так:

```stylus
&__elem {
    background url(s('%s__image.png', $ctx))
}
```

**Comment by just-boris** [comment]
@h4 у меня less, поэтому получилось как-то так:
```
.row {
  @name: .row;
  &__controls { 
    visibility: hidden;
    @{name}_hovered & {
      visibility: visible;
    }
  }
}
```

в случае, когда стилей больше чем пара строк, лучше сохранить имя блока в переменную, чтобы не потерялось при рефакторингах

**Comment by h4** [comment]
@just-boris переходи на Stylus. Тут можно не писать скобочки и точки с запятой ;)

**Comment by sipayRT** [comment]
@h4 интересный кейс, спс

**Comment by just-boris** [comment]
@h4 зато нельзя пользоваться mixin-ами bootstrap и переопределять его переменные.
Но это уже другая история.

**Comment by just-boris** [comment]
Всем спасибо за показанные примеры!

**Comment by zxqfox** [comment]
Вообще, `{$ctx}` штука полезная и для переименования блоков, поэтому точка там противопоказана. Ну и в целом лучше `{$block}` или `{$b}`, чтобы была возможность еще и `{$elem}` или `{$e}` использовать.

**Comment by belozyorcev** [comment]
@zxqfox +1. Будто с моей головы написал слова ;)

**Comment by h4** [comment]
@zxqfox ну я `{$ctx}` тупо из кода *bem-components* позаимствовал.
